### PR TITLE
handle cull mode by indices, refer to issue #10172

### PR DESCRIPTION
**Pull request by weihuoya**
improve performance for GVGNP.

See #10172

### CODE REVIEW

**Comment by hrydgard** [comment]
Nice, this is exactly the kind of thing I had imagined when I wrote that comment :)

Looks like it deserves some testing and, if I'm understanding this right, it should be done for the other triangle primitive types as well (triangle list, fan, plus all the non-indexed types). Could also fall back to flushing the cull mode if those are encountered...

**Comment by weihuoya** [comment]
I haven't find a game to test other primitive types, so all those fallback to flush.

**Comment by weihuoya** [comment]
GTA use triangle fan, need handle it.

File: GPU/Common/IndexGenerator.cpp [hunk 1]
```diff
@@ -253,11 +255,13 @@ void IndexGenerator::TranslateStrip(int numInds, const ITypeLE *inds, int indexO
...
 
 template <class ITypeLE, int flag>
-void IndexGenerator::TranslateFan(int numInds, const ITypeLE *inds, int indexOffset) {
+void IndexGenerator::TranslateFan(int numInds, const ITypeLE *inds, int indexOffset, bool clockwise) {
 	if (numInds <= 0) return;
 	indexOffset = index_ - indexOffset;
 	int numTris = numInds - 2;
 	u16 *outInds = inds_;
+	const int v1 = clockwise ? 1 : 2;
+	const int v2 = clockwise ? 2 : 1;
```
Comment (unknownbrackets) [inline GPU/Common/IndexGenerator.cpp]: Didn't use these below?

-[Unknown]

Comment (weihuoya) [reply GPU/Common/IndexGenerator.cpp]: I will fx it now.:flushed:

File: GPU/GPUCommon.cpp [hunk 2]
```diff
@@ -1618,6 +1632,14 @@ void GPUCommon::Execute_Prim(u32 op, u32 diff) {
 			gstate.cmdmem[GE_CMD_TEXSCALEV] = data;
 			gstate_c.uv.vScale = getFloat24(data);
 			break;
+		case GE_CMD_TEXOFFSETU:
```
Comment (weihuoya) [inline GPU/GPUCommon.cpp]: Tomb Raider use GE_CMD_TEXOFFSETU and GE_CMD_TEXOFFSETV frequently, but not quite useful.:grimacing:

Comment (unknownbrackets) [reply GPU/GPUCommon.cpp]: I think it's some engine, there's other games that do that too (basically the whole point of the `gstate_c.uv` optimization.)  Makes sense to add them here.

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]
Looks good to me - I think we should merge.

`dc.cullMode != -1 && gstate.isCullEnabled() && gstate.getCullMode() != dc.cullMode` could possibly be factored into a method, but I think it's a good change either way.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, agreed.

**Comment by Jinzombie** [comment]
WOW 😃 Naruto Ultimate Ninja Impact 1 now running smooth in 30/30 FPS unlike in previous version this game is running very slow but now it's fast maybe because of this "handle cull mode of indices"?

**Comment by Jinzombie** [comment]
I will try Digimon Adventure if this PR help its #10842

**Pull request merged**